Rewards auto contribution support

emerick opened the pull request:
Resolves https://github.com/brave/brave-browser/issues/12107
Resolves https://github.com/brave/brave-browser/issues/12781

Add auto-contribution support for Rewards media publishers.

emerick commented:
Reviewers: Please note that this PR is based off the `rewards-greaselion-brave-extension` branch, so it's probably simplest to review by commit and simply ignore that particular commit.

marshall commented on scripts/brave_rewards/publisher/common/mediaDuration.ts:
```diff
@@ -0,0 +1,36 @@
+let firstVisit = true
+
+export const setFirstVisit = (enabled: boolean) => {
+  firstVisit = enabled
```
should `firstVisit` be also persisted for each site? I may not fully understand it's purpose, but I assume the flag will get reset whenever Brave is restarted, since it's only in memory.

emerick replied:
It's only relevant for the length of the browser session, but you just made me realize that it's not going to work as expected now that I made this common code. Thanks for bringing it up, will fix.

marshall commented on scripts/brave_rewards/publisher/reddit/redditAutoContribution.ts:
```diff
@@ -0,0 +1,49 @@
+import * as types from './types'
+import * as utils from './utils'
+
+let lastActivityTime = 0
+
+const resetActivityTime = () => {
+  lastActivityTime = Date.now()
+}
+
+const initScript = () => {
+  // Don't run in incognito context
+  if (chrome.extension.inIncognitoContext) {
+    return
+  }
+
+  createPort((success: boolean) => {
+    if (!success) {
+      console.error('Failed to initialize communications port')
+      return
+    }
+
+    // Calculate duration on visibility change
+    document.addEventListener('visibilitychange', function () {
+      if (document.visibilityState === 'visible') {
+        resetActivityTime()
+      } else {
```
this code block (and lines 38-49) seem to be pretty common between scripts. since they're pretty crucial to counting page activity, do you think it makes sense to factor out the code and put it in `common`?

emerick replied:
Yes, it's been bugging me too. Let me create an issue to refactor this, because I want to do it for all of the scripts (base, publisherInfo, and autoContribution) in the various Rewards media publishers but I think that will get a bit big for this PR.

emerick replied:
https://github.com/brave/brave-browser/issues/12698

marshall reviewed (approved):
looks great. there is a pretty big chunk of common logic between all of the auto contribution scripts you might consider refactoring, but I don't think it needs to hold up this PR.

zenparsing commented on scripts/brave_rewards/publisher/vimeo/publisherInfo.ts:
```diff
@@ -182,3 +176,11 @@ export const send = () => {
   }
 }
+
+export const get = async () => {
```
I see you're gravitating toward short exported names. Nothing wrong with that, although from a practical standpoint it means that importers need to use `import * as` namespace imports instead of pulling in the names that they need.

emerick replied:
Yeah, that's a good point. I made that change because `publisherInfo.getPublisherInfo` seemed excessive, maybe I got too carried away. It seems like either choice will be a little problematic. I think I'll keep it as-is for now, but will give it some more thought for the future.

zenparsing commented on scripts/brave_rewards/publisher/youtube/mediaDuration.ts:
```diff
@@ -13,11 +15,11 @@ export const setFirstVisit = (enabled: boolean) => {
 }
 
-export const sendMetadata = (url: URL) => {
+export const sendMetadataFromUrl = (url: URL) => {
```
Is "from" intentional? Should it be `sendMetadataForUrl`?

emerick replied:
Intentional, but maybe not obvious. I was trying to signify that the metadata is coming from the URL (i.e., we parse it out of the URL).

zenparsing commented on scripts/brave_rewards/publisher/common/utils.ts:
```diff
@@ -2,6 +2,23 @@
+  }
+
+  // This clause was needed for Vimeo
+  if (mediaId.startsWith(mediaType)) {
```
Seems like this could have surprising results for other publisher types, e.g. `twitter` is a valid media ID for the `twitter` media type. Can this special-case logic be moved into the vimeo modules somehow?

emerick replied:
Yeah, I'll do that. I'm not even sure that this code is strictly necessary, it just seemed weird to have a media ID of `vimeo_vimeo_12345`.

emerick replied:
This turned out to be a bug (I was passing the media key instead of the raw media ID). Fixed now, thanks for mentioning it.

zenparsing reviewed (commented):
General question: is media duration tracked appropriately for sites that use the history API? I didn't see "updatedTab" listeners in the new code.

emerick commented:
@zenparsing Addressed the issue you mentioned about using the History API. Should be ready for review again, thanks.

marshall reviewed (changes requested):
overall looks great. I left a few nits but they aren't blockers IMO.

I also pulled this down to test locally and ran into a few minor issues:

* The favicon for github profiles seems to be blank (I tested with my profile by visiting it directly)
* Visiting a youtube channel page seems to count towards AC for youtube.com, instead of a specific profile, and I couldn't get videos to count towards either youtube.com or a specific profile. (The channel I tested with was shared privately)

I think once these are addressed this is good to land. Also it looks like this needs a fresh rebase

emerick commented on scripts/brave_rewards/publisher/common/mediaDuration.ts:
```diff
@@ -0,0 +1,28 @@
+
+  port.postMessage({
+    type: 'MediaDurationMetadata',
+    mediaType: mediaType,
```
Fixed.

emerick commented on scripts/brave_rewards/publisher/common/mediaDuration.ts:
```diff
@@ -0,0 +1,28 @@
+
+import * as utils from '../common/utils'
+
+export const sendMetadata = (mediaType: string, mediaId: string, duration: number, firstVisit: boolean) => {
```
Fixed. Yes, makes sense - I'll enter an issue for that.

emerick replied:
https://github.com/brave/brave-browser/issues/12874

emerick commented:
@marshall Addressed all feedback, though I haven't been able to repro the missing Github favicon. Maybe try with a clean profile? The Youtube issue was an actual bug; we needed to use the YouTube channel ID as the media ID when there's no media key.

marshall commented on scripts/brave_rewards/publisher/youtube/publisherInfo.ts:
```diff
@@ -2,161 +2,157 @@
+  if (!response.ok) {
+    if (response.status === 401) {
+      // Embedding disabled; need to scrape data from page instead
+      return scrapePublisherInfoFromPage(url)
```
should this be `return await scrapePublisherInfoFromPage(url)` since the funciton is async ?

zenparsing replied:
The value returned from an async function is implicitly awaited (technically, the promise corresponding to the async function invocation is resolved with the return value, and resolution will implicitly unwrap promises).

emerick replied:
Yeah and actually tslint spits out the following if you attempt it:

```
C:/work/brave-site-specific-scripts/scripts/brave_rewards/publisher/youtube/publisherInfo.ts:98:14
ERROR: 98:14  no-return-await  Unnecessary 'await'.
```

marshall replied:
awesome, I wasn't aware of this behavior. thanks for clarification!

marshall commented on scripts/brave_rewards/publisher/youtube/publisherInfo.ts:
```diff
@@ -166,110 +162,70 @@ const sendForVideoHelper = (url: string, responseText: string, publisherName: st
   if (utils.isUserPath(location.pathname)) {
-    sendForUser()
-    return
+    return getPublisherInfoForUser()
```
shouldn't these also be be `await`ed?

emerick replied:
Our current tslint settings complain about it being "unnecessary" (see above).

marshall replied:
cool thanks for clarifying!

marshall reviewed (commented):
@emerick looks like there is some async code not using await.. let me know if they're expected or not, and I can test once it's addressed

emerick commented:
This fix will be in Greaselion `1.0.39`.
